CAR random variables

aerubanov opened the pull request:
close #4518 
I Changed CAR distribution class to compatible with ```v4``` ```RandomVariable``` and implemented CAR random variable class. But now I need some advises for further improvements. 

 @brandonwillard, I have few questions about random variable for CAR:
- if some parameter of random variable have param that may be scalar or vector (like ```mu```, ```tau``` and ```alpa```), what I should specify in ```ndims_params```? If I pass scalar instead vector  an error occurred.
- Should I somehow handle size param in ```rng_fn``` method?
- Where I should store intermediate variables (like ```D``` and ```lam```) that need to be calculated only once.  Initially they were calculated in the ```__init__``` method of distribution class and stored as distribution class atribute.

@ckrapu, Do you have any ideas for testing the implementation of the algorithm?

aerubanov committed:
add CAR random variable and change CAR distribution

aerubanov committed:
Add CAR RandomVariable Op and refactor CAR distribution

aerubanov committed:
fix formating

brandonwillard commented:
> * if some parameter of random variable have param that may be scalar or vector (like `mu`, `tau` and `alpa`), what I should specify in `ndims_params`? If I pass scalar instead vector  an error occurred.

It sounds like you need to find a "minimally sufficient" space into which you can embed all the parameters.  The real question is "What's the correspondence between the dimensions of these parameters?"  For instance, if `mu` is a scalar, what should `tau` be, etc.?


> * Should I somehow handle size param in `rng_fn` method?

Yes, that's a mandatory parameter.  You'll need to understand those dimensions above in order to implement this, though; otherwise, take a look at the implementation of [`MvNormalRV`](https://github.com/pymc-devs/aesara/blob/02378861f1a77135f2556018630092a09262ea76/aesara/tensor/random/basic.py#L146) for a brute-force approach to implementing `size`.

> * Where I should store intermediate variables (like `D` and `lam`) that need to be calculated only once.  Initially they were calculated in the `__init__` method of distribution class and stored as distribution class atribute.

You can add those as parameters to the `Op` and use `Op.__init__` to compute the derived quantities and hide them from the caller.

aerubanov commented:
> It sounds like you need to find a "minimally sufficient" space into which you can embed all the parameters. The real question is "What's the correspondence between the dimensions of these parameters?" For instance, if `mu` is a scalar, what should `tau` be, etc.?

@brandonwillard No, my question was about another thing.  The CAR class docs specifying that a float or array can be passed as these parameters, optionaly. How I should I handle this? Should I transform float to array in ```dist``` method? Or some other way?

brandonwillard commented:
> Should I transform float to array in `dist` method? Or some other way?

If you have the parameter dimensions worked out already, then there's no question about whether or not you should&mdash;for instance&mdash;convert a scalar to a vector, matrix, etc.  If it's a question of _where_ to do the conversion, often that kind of thing is done in `Op.make_node`.

Assuming that's the case, and you need to convert a scalar to a vector, just add a broadcastable dimension to the scalar value.  There are `aesara.tensor.shape_pad*` functions that will do this, as well as `aesara.tensor.reshape`.

(Aesara could probably use implementations of the `numpy.atleast_*`, too.)

N.B.: There are also some broadcasting helper functions used throughout the `RandomVariable` implementations in Aesara.  For instance, check out [`broadcast_params`](https://github.com/pymc-devs/aesara/blob/02378861f1a77135f2556018630092a09262ea76/aesara/tensor/random/utils.py#L53).

brandonwillard commented on pymc3/distributions/multivariate.py:
```diff
@@ -1940,6 +1940,44 @@ def _distr_parameters_for_repr(self):
+    _print_name = ("CAR", "\\operatorname{CAR}")
+
+    @classmethod
+    def rng_fn(cls, rng: np.random.RandomState,  mu, W, alpha, tau, size):
```
This method looks like it's assuming that the `RandomVariable` has only 4 parameters, but the `ndims_params` above says it has 5.

aerubanov replied:
Oops, my mistake. I originally planned to pass ```sparse``` to ```RandomVariable```, but then I abandoned that and didn't change ```ndims_params```. I'll fix that.

aerubanov committed:
fix ndims_params

aerubanov commented:
@brandonwillard , Thanks for the tips and quick response! I will make the necessary changes.

ckrapu commented:
> @ckrapu, Do you have any ideas for testing the implementation of the algorithm?


The testing strategy I used for the original CAR implementation was to make sure that the CAR logp and the equivalent multivariate normal with CAR-structured covariance matrix were equivalent, up to an additive constant. You can see how that is implemented [here](https://github.com/pymc-devs/pymc3/blob/2dee9dafb3541043ae23b66fe9262e70fbd2769a/pymc3/tests/test_distributions.py#L2619).

ckrapu commented on pymc3/distributions/multivariate.py:
```diff
@@ -1940,6 +1940,44 @@ def _distr_parameters_for_repr(self):
+            u += 1
+        L = scipy.linalg.cholesky_banded(Qb, lower=False)
+        z = rng.normal(size=W.shape[0], loc=mu)
+        samples = scipy.linalg.cho_solve_banded((L, False), z)
```
I'm not familiar with the desired signature of `rng_fn`, but it's not clear to me that this allows for sampling multiple CAR draws simultaneously. If that's not the case, then one-time solution to the permutation optimization problem in `reverse_cuthill_mckee` is not going to be better than simply sampling from a multivariate normal with the CAR covariance matrix.

aerubanov replied:
I think that the size parameter is just what you need to implement the possibility of sampling multiple CAR draws simultaneously. So I should change my implementation of this method.

aerubanov committed:
fix formating

aerubanov committed:
add make_node

brandonwillard commented on pymc3/distributions/multivariate.py:
```diff
@@ -1940,6 +1940,53 @@ def _distr_parameters_for_repr(self):
+    dtype = "floatX"
+    _print_name = ("CAR", "\\operatorname{CAR}")
+
+    def make_node(self, rng, size, dtype, *dist_params):
```
It's better to explicitly specify the parameters (i.e. `mu`, `W`, etc.); otherwise, the signature is unnecessarily ambiguous.  

Also, you'll likely need to convert those distribution parameters to Aesara variables using `aesara.tensor.as_tensor[_variable]`.

aerubanov replied:
@brandonwillard, Should I do this convertion in the ```make_node``` method of ```RandomVareable``` instead of the ```dist``` method of CAR ```Distribution``` class?

brandonwillard replied:
Yeah, it's always good to do that in `Op.make_node`.

aerubanov committed:
move convertion to at in make_node

aerubanov committed:
add size support in rng_fn

aerubanov committed:
change make node

aerubanov committed:
fix typo

aerubanov commented:
I add handling of ```size``` param in ```rng_fn``` method and move convertion to aesara tensor in ```make_node``` method. But now, when it trying to calculate ```logp``` for CAR distribution i get this error:
```puython
n[2]: import pymc3 as pm, numpy as np
Backend TkAgg is interactive backend. Turning interactive mode on.
In[3]: W = np.array([[1, 0, 0, 1, 0], [0, 1, 0, 1, 0], [0, 0, 1, 0, 1], [1, 1, 0, 1, 0], [0, 0, 1, 0, 1]])
In[4]: pm.logpt(pm.CAR.dist([0, 0, 0, 0, 0], W, [0.5, 0.5, 0.5, 0,5], [1, 1, 1, 1, 1]),np.random.randn(5)).eval()
Traceback (most recent call last):
  File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3437, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-71c3b392bce9>", line 1, in <module>
    pm.logpt(pm.CAR.dist([0, 0, 0, 0, 0], W, [0.5, 0.5, 0.5, 0,5], [1, 1, 1, 1, 1]),np.random.randn(5)).eval()
  File "/home/anatoly/HDD/Projects/pymc3/pymc3/distributions/logp.py", line 165, in logpt
    rv_value = rv_var.type.filter_variable(rv_value.astype(rv_var.dtype))
  File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/type.py", line 258, in filter_variable
    raise TypeError(
TypeError: Cannot convert Type TensorType(float64, vector) (of Variable TensorConstant{[ 2.000324...65305157]}) into Type TensorType(float64, matrix). You can try to manually convert TensorConstant{[ 2.000324...65305157]} into a TensorType(float64, matrix).
```
I guess I just did something wrong on the last call. @brandonwillard , could you help me with these error?

I also plan to start figuring out how to write tests.Based on the discussion above, I think we can compare the sampling from CAR with the sampling from the equivalent multivariate normal with CAR-structured covariance matrix. But perhaps it is not the samples themselves that should be compared, but the statistics calculated from their distribution or something similar.

ckrapu commented:
```
> TypeError: Cannot convert Type TensorType(float64, vector) (of Variable TensorConstant{[ 2.000324...65305157]}) into Type TensorType(float64, matrix). You can try to manually convert TensorConstant{[ 2.000324...65305157]} into a TensorType(float64, matrix).
 ```

If you try using `np.random.randn(5,1)` instead, does it work?

brandonwillard commented on pymc3/distributions/multivariate.py:
```diff
@@ -1940,6 +1943,75 @@ def _distr_parameters_for_repr(self):
+class CARRV(RandomVariable):
+    name = "car"
+    ndim_supp = 1
+    ndims_params = [1, 2, 1, 1]
```
Are the last two dimensions really supposed to be vectors?  Below it states that they can be scalars.

brandonwillard commented on pymc3/distributions/multivariate.py:
```diff
@@ -1940,6 +1943,75 @@ def _distr_parameters_for_repr(self):
+        if tau.ndim == 0:
+            tau = tau * at.ones(mu.shape)
+        if alpha.ndim == 0:
+            alpha = alpha * at.ones(mu.shape)
```
If `alpha` and `tau` are supposed to be scalars, this will need to be removed.

Regardless, it looks like these steps are trying to broadcast `tau` and `alpha`, but this kind of broadcasting is costly, because new arrays will be allocated.  If you want to broadcast, use something like `at.broadcast_to` (or `at.extra_ops.broadcast_to` for earlier versions of Aesara).

aerubanov replied:
@brandonwillard , When I wrote this code, I was guided by the original implementation of CAR distribution from PR #4504 where alpha and tau may be scalar or vector, optionally.  I was trying to keep this opportunity alive. As I understand it, this feature is only needed to reduce the amount of user code, so in principle we can make the parameters alpha and tau can only be a vector.
@ckrapu , could you comment this?

aerubanov replied:
To clarify: as I understand it, alpha and tau are vectors, but often all their elements are the same. So being able to specify a scalar instead of a vector may be useful.

brandonwillard commented on pymc3/distributions/multivariate.py:
```diff
@@ -1940,6 +1943,75 @@ def _distr_parameters_for_repr(self):
+            tau = tau * at.ones(mu.shape)
+        if alpha.ndim == 0:
+            alpha = alpha * at.ones(mu.shape)
+        return super().make_node(rng, size, dtype, *(mu, W, alpha, tau))
```
There's no need for the `*` here.

brandonwillard commented on pymc3/distributions/multivariate.py:
```diff
@@ -1981,45 +2053,17 @@ class CAR(Continuous):
+    @classmethod
+    def dist(cls, mu, W, alpha, tau, sparse=False, *args, **kwargs):
 
         if not W.ndim == 2 or not np.allclose(W, W.T):
```
NumPy cannot be used here, because `W` could be a `TensorVariable`.  You'll need to assert symmetry in a symbolic way&mdash;or simply remove the check for now.

brandonwillard replied:
@aerubanov, this still needs to be resolved.

aerubanov replied:
@brandonwillard , Thanks for the reminder, I made the changes.

brandonwillard commented on pymc3/distributions/multivariate.py:
```diff
@@ -1981,45 +2053,17 @@ class CAR(Continuous):
+        return super().dist([mu, W, alpha, tau, sparse], **kwargs)
 
-    def logp(self, value):
+    def logp(value, mu, W, alpha, tau, sparse=False):
```
There's no need for a `sparse` parameter here; plus, this option would only be useful to someone directly calling this function, which isn't too likely or common.  

More importantly, you can determine whether or not `W` is sparse by inspection.

brandonwillard commented on pymc3/distributions/multivariate.py:
```diff
@@ -2036,30 +2080,34 @@ def logp(self, value):
+
+        Dinv_sqrt = np.diag(1 / np.sqrt(D))
+        DWD = np.matmul(np.matmul(Dinv_sqrt, W), Dinv_sqrt)
+        lam = scipy.linalg.eigvalsh(DWD)
```
NumPy can't be used here; all these variables are symbolic (i.e. `TensorVariable`s).  You'll need to use their Aesara equivalents.

brandonwillard commented on pymc3/distributions/multivariate.py:
```diff
@@ -2036,30 +2080,34 @@ def logp(self, value):
-        raise NotImplementedError("Sampling from a CAR distribution is not supported.")
-
     def _distr_parameters_for_repr(self):
         return ["mu", "W", "alpha", "tau"]
```
This can be removed; it's no longer necessary.

brandonwillard commented:
I pushed some of the recommendations I made above and re-enabled the old CAR test in `test_distributions`, so you'll need to rebase locally.

Otherwise, according to the old test, it looks like the log-likelihood could be working; however, that test needs an addition for at least one `size` > 2D.  

Also, it looks like we need to add a CAR test to `test_distributions_random` for the random sampling in `CARRV` (I couldn't find one in there).

aerubanov commented:
Thanks for review, @brandonwillard !
> Otherwise, according to the old test, it looks like the log-likelihood could be working; however, that test needs an addition for at least one `size` > 2D.

Yes, I will try to add test some tests for  such case.


> Also, it looks like we need to add a CAR test to `test_distributions_random` for the random sampling in `CARRV` (I couldn't find one in there).

I think about solution based on comparison samples from ```CAR``` and ```MultivariateNormal``` with CAR-structured covariance matrix, similar to what we do for ```logpt```. I discused this idea with @ferrine , and he suggested to use Sliced-Wasserstein distance for comparison metric . What do you think?

ricardoV94 commented:
> I think about solution based on comparison samples from `CAR` and `MultivariateNormal` with CAR-structured covariance matrix, similar to what we do for `logpt`. I discused this idea with @ferrine , and he suggested to use Sliced-Wasserstein distance for comparison metric . What do you think?

We have some tests for random methods that use the Kolgomorov-Smirnov test to assess the equivalence between pymc3 and a reference number generator (which could be the MultivariateNormal in your case). Maybe you can adapt the logic from there: https://github.com/pymc-devs/pymc3/blob/e5c42b47e43940837afe171f7a30c8ea89f54ed2/pymc3/tests/test_distributions_random.py#L60

aerubanov commented:
@ricardoV94 , thank you, I will check this.

aerubanov committed:
add 3d test for CAR logpt

aerubanov committed:
add test for CARRV rng_fn()

aerubanov commented:
I added a test for the ```rng_fn``` method based on the logic from the ```pymc3_random()``` function in ```pymc3/pymc3/tests/test_distributions_random.py```, but it fails =(. Now I'm trying to fix it.

@brandonwillard , @ricardoV94, do you have any ideas what I am doing wrong?

ricardoV94 commented:
Does the test pass if you duplicate the CAR and compare against itself, instead of against the MvNormal? If it does it means the two distributions you have now are not really equivalent

aerubanov commented:
@ricardoV94 , yes, if I duplicate CAR distribution, the test passes. But as @ckrapu pointed out, this test is not correct for this distribution. So I will try to use the options he suggested.

aerubanov commented:
I experimented with different W matrix formats, and sometimes the test passed, but sometimes failed. So, it seems that the KS test is actually not appropriate here.

brandonwillard commented:
> I experimented with different W matrix formats, and sometimes the test passed, but sometimes failed. So, it seems that the KS test is actually not appropriate here.

If they're only failing within a smallish window near the numerical limit/precision, then that's fine.  Also, we'll need to seed the tests, which will obviously help.

aerubanov committed:
change test_car_rng_fn

aerubanov commented:
I changed test for CAR ```rng_fn``` method. Now we compare samples for each dimensions piece by piece. and looks like it work better. @brandonwillard,

> Also, we'll need to seed the tests, which will obviously help.

I added numpy random seed setting before starting to generate samples. Is there anything else that needs to be added?

aerubanov committed:
change seed setting and rebase

aerubanov committed:
move symmetry check to make_node

aerubanov committed:
remove numpy usage from dist method

brandonwillard commented on pymc3/distributions/multivariate.py:
```diff
@@ -1940,6 +1943,73 @@ def _distr_parameters_for_repr(self):
+
+        if sparse:
+            W_sparse = scipy.sparse.csr_matrix(W)
+            W = aesara.sparse.as_sparse_variable(W_sparse)
```
```suggestion
            W = aesara.sparse.as_sparse_variable(W)
```

Just like before, `W` could be a `TensorVariable`, so calling `scipy.sparse.csr_matrix` isn't sound.

Also, as I mentioned in the other comment, the `sparse` option is unnecessary; the caller should provide a sparse variable if they want to use one, and the logic here should handle both cases.

aerubanov committed:
remove sparce argument

aerubanov committed:
fix logp

aerubanov commented:
@brandonwillard , I remove ```sparse``` argument, but now I have two problems:
- when I try to pass aesara ```SparseVariable``` as ```W``` in ```CAR``` constructor, I get this error:
  ```
   File "<ipython-input-13-b44074052d76>", line 2, in <module>
    car = pm.CAR("car", mu, ws, alpha, tau)
  ..........................
  File "/home/anatoly/HDD/Projects/pymc3/pymc3/distributions/multivariate.py", line 1963, in make_node
    return super().make_node(rng, size, dtype, mu, W, alpha, tau)
  File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py", line 374, in make_node
    bcast = self.compute_bcast(dist_params, size)
  File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/configparser.py", line 49, in res
    return f(*args, **kwargs)
  File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py", line 287, in compute_bcast
    shape = self._infer_shape(size, dist_params)
  File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py", line 224, in _infer_shape
    params_ind_slice = tuple(
  File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py", line 225, in <genexpr>
    slice_ind_dims(p, ps, n)
  File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py", line 214, in slice_ind_dims
    for s, b in zip(shape[:-n], p.broadcastable[:-n])
  AttributeError: 'SparseVariable' object has no attribute 'broadcastable'
  ```
  But if I pass ```ndarray``` or ```TensorVariable``` all works fine.
- test for ```CAR``` ```logp``` method fails for 3D ```size```, because ```aesara.sparse.dot()``` accept only 1 or 2 ```ndim``` inputs.

What should I do about these errors?

brandonwillard commented:
> * when I try to pass aesara `SparseVariable` as `W` in `CAR` constructor, I get this error:

Add a test for this case and I'll take a look.

> * test for `CAR` `logp` method fails for 3D `size`, because `aesara.sparse.dot()` accept only 1 or 2 `ndim` inputs.

Sounds like a condition is needed for `SparseVariable`s/non-`SparseVariable`s.

aerubanov committed:
add tests for sparse W

aerubanov commented:
@brandonwillard , I added tests for case when ```W``` is sparse variable, see ```test_car_logp``` and ```test_car_rng_fn``` in ```pymc3/tests/test_distributions.py```.

ricardoV94 commented:
@aerubanov It seems your new tests are failing. Did you figure this out in the meantime?

Edit: Please not that the branch `v4` has been merged to `main`, you should change the target of your PR accordingly :)

aerubanov commented:
@ricardoV94 , I described a problem with sparse variable few messages above and write tests that reproduce it. But error take place  in the `make_node()` method of `RandomVariable` itself, not in my code, and I not very familiar with it. So I asked @brandonwillard  to take a look. 

P.s.: change PR target.

aerubanov commented:
@ricardoV94 , After changing PR target I get file conflicts but github do not allow me to resolve it without write acces. Should I do it in some another way or it`s OK?

brandonwillard commented:
@aerubanov, I rebased your PR onto `main`; that should remove all the unrelated commits and bring it up to date.  You'll need to re-check it out locally, or pull in the changes.

brandonwillard reviewed (commented):
> * when I try to pass aesara `SparseVariable` as `W` in `CAR` constructor, I get this error:
>   ```
>    File "<ipython-input-13-b44074052d76>", line 2, in <module>
>     car = pm.CAR("car", mu, ws, alpha, tau)
>   ..........................
>   File "/home/anatoly/HDD/Projects/pymc3/pymc3/distributions/multivariate.py", line 1963, in make_node
>     return super().make_node(rng, size, dtype, mu, W, alpha, tau)
>   File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py", line 374, in make_node
>     bcast = self.compute_bcast(dist_params, size)
>   File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/configparser.py", line 49, in res
>     return f(*args, **kwargs)
>   File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py", line 287, in compute_bcast
>     shape = self._infer_shape(size, dist_params)
>   File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py", line 224, in _infer_shape
>     params_ind_slice = tuple(
>   File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py", line 225, in <genexpr>
>     slice_ind_dims(p, ps, n)
>   File "/home/anatoly/anaconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py", line 214, in slice_ind_dims
>     for s, b in zip(shape[:-n], p.broadcastable[:-n])
>   AttributeError: 'SparseVariable' object has no attribute 'broadcastable'
>   ```
>   
>   
>     
>       
>       
>   
>       
>       
>   
>     
>   
>   
>   But if I pass `ndarray` or `TensorVariable` all works fine.

The issue you're seeing is that sparse variables don't follow the same design as regular/dense variables.  What the error is saying is that `SparseVariable` (and, equivalently, `SparseType`) doesn't have a `broadcastable` property, like `TensorVariable`/`TensorType` does, and the helper function in `RandomVariable._infer_shape` uses that property.

This is being addressed in generality by https://github.com/aesara-devs/aesara/pull/303, but there are much more direct ways to handle this.  The easiest might involve a change to `RandomVariable._infer_shape`, but your custom `RandomVariable` could also override `_infer_shape` and avoid this step when the inputs are sparse.

twiecki commented:
@aerubanov Any progress on this? Would be great to get this merged.

aerubanov commented:
@twiecki ,I think I'll get it done in the next couple of days.

aerubanov committed:
add _infer_shape()

aerubanov commented:
I added the `_infer_shape()` method and that solved the "'SparseVariable' object has no attribute 'broadcastable'" error. But now I have a different problem in the `logp` method, because the `.sum()` method is not implemented for the sparse variable. Now I'm trying to find a workaround.

aerubanov committed:
walkaround for sparse .sum()

aerubanov commented on pymc3/distributions/multivariate.py:
```diff
@@ -2036,30 +2074,41 @@ def logp(self, value):
+            D = [
+                aesara.sparse.basic.csm_data(W[i : i + 1, :]).eval().sum()
+                for i in range(int(W.shape[0].eval()))
+            ]
```
@brandonwillard , I added this as a workaround for the `.sum()` method, which is not implemented for a sparse variable. But maybe there is a more optimal solution?

ricardoV94 replied:
@brandonwillard is this something that should be implemented at the aesara level?

brandonwillard replied:
Absolutely

twiecki replied:
Is there an issue already?

aerubanov commented:
@brandonwillard, I just want to remind you of my question above. I hope I'm not distracting you too much.

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -2933,6 +2942,48 @@ def test_car_logp(size):
 
 
+@pytest.mark.parametrize("sparse", [True, False])
+def test_car_rng_fn(sparse):
```
This would probably be better placed in `test_distributions_random.py`

aerubanov replied:
I moved it into `test_distributions_random.py`.

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -2933,6 +2942,48 @@ def test_car_logp(size):
+    p, f = delta, n_fails
+    while p <= delta and f > 0:
+        car_smp, mn_smp = check["car"][f - 1, :, :], check["mn"][f - 1, :, :]
+        p = min(
```
What is the rationale for using the `min`, is this not making the test easier than the assumed `delta`?

ricardoV94 replied:
Nevermind, I read it the wrong way around. `min` is the strongest test...

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -2933,6 +2942,48 @@ def test_car_logp(size):
+@pytest.mark.parametrize("sparse", [True, False])
+def test_car_rng_fn(sparse):
+    delta = 0.05  # limit for KS p-value
+    n_fails = 100  # Allows the KS fails a certain number of times
```
Why such a large `n_fails`?

aerubanov replied:
Becouse I forgot to change it back after experimenting with it. Thank you, I fixed it.

aerubanov committed:
move test_car_rng_fn()

aerubanov commented:
@ricardoV94,I made some changes based on your comments, could you look it up?

ricardoV94 commented on pymc3/distributions/multivariate.py:
```diff
@@ -12,7 +12,7 @@
 #   limitations under the License.
 
-#!/usr/bin/env python
+# !/usr/bin/env python
```
```suggestion
#!/usr/bin/env python
```

ricardoV94 commented on pymc3/distributions/multivariate.py:
```diff
@@ -1940,6 +1943,73 @@ def _distr_parameters_for_repr(self):
+    _print_name = ("CAR", "\\operatorname{CAR}")
+
+    def make_node(self, rng, size, dtype, mu, W, alpha, tau):
+        mu = at.as_tensor_variable(mu)
```
Should these be of type `mu = at.as_tensor_variable(floatX(mu))` and so on for the other variables?

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -2896,34 +2896,43 @@ def test_orderedlogistic_dimensions(shape):
+        W = aesara.sparse.csr_from_dense(W)
+
+    car_dist = CAR.dist(mu, W, alpha, tau, size=size)
+    # xs = np.broadcast_to(xs, size + mu.shape)
```
```suggestion
```

ricardoV94 reviewed (commented):
Sorry for taking so long to review. I left some minor comments/ questions. Otherwise, it seems like this PR is almost over the finish line!

Also you need to rebase from main to solve the conflict issues

aerubanov committed:
Update pymc3/distributions/multivariate.py

Co-authored-by: Ricardo Vieira <28983449+ricardoV94@users.noreply.github.com>

aerubanov committed:
Update pymc3/tests/test_distributions.py

Co-authored-by: Ricardo Vieira <28983449+ricardoV94@users.noreply.github.com>

aerubanov committed:
add floatx conversion

ricardoV94 commented:
@aerubanov you need to rebase from main and fix the conflict issues. Also, would you mind opening an issue on Aesara about the missing sum for sparse variables (and a reminder issue here to fix it once that's solved)?

@brandonwillard Is it okay to merge with the current `.eval()` hack or should we wait for an upstream fix / find another alternative if possible?

brandonwillard commented:
> @brandonwillard Is it okay to merge with the current `.eval()` hack or should we wait for an upstream fix / find another alternative if possible?

No, we shouldn't merge evaluations like that.

aerubanov commented:
@ricardoV94 I rebased on ```main``` and I will open issues about ```.sum()``` for sparse tensor in aesara and here.

aerubanov committed:
remove hack for .sum()

aerubanov commented:
@ricardoV94, I removed  ```.eval()``` hack. @brandonwillard noticed me in https://github.com/aesara-devs/aesara/issues/522 that ```sp_sum()``` fit mach better here.

aerubanov committed:
fix formating

aerubanov committed:
add tolerance setting based on float type

aerubanov commented:
@ricardoV94, I changed formatting and add tolerance selection based on the ```aesara.config.floatX``` value in ```test_car_logp```. These should fixed failed tests.

ricardoV94 commented on pymc3/distributions/multivariate.py:
```diff
@@ -1925,6 +1924,73 @@ def _distr_parameters_for_repr(self):
+        mu = at.as_tensor_variable(floatX(mu))
+
+        W = aesara.sparse.as_sparse_or_tensor_variable(floatX(W))
+        if not W.ndim == 2:
```
This check seems less restrictive than the original one, and it only covers the number of dimensions. Any reason for the change? You can add a `symbolic` assert if the point was to allow for symbolic matrices. 

Also, it seems you can move these checks and `as_tensor_variable` calls to the `dist` classmethod, which is a bit more in line with how the other distributions in the library are set.

aerubanov replied:
@ricardoV94 , I removed symmetry check because I can not find an implementation of  ```all_close``` for aesara sparse matrices. I only found these one https://aesara.readthedocs.io/en/latest/library/tensor/basic.html#aesara.tensor.allclose for dense. 

In the discussion above (https://github.com/pymc-devs/pymc3/pull/4596#discussion_r606376156), @brandonwillard said that ```as_tensor_variable``` calls should be placed in the ```make_node``` method rather than ```dist```. And I have done that.

ricardoV94 replied:
> @ricardoV94 , I removed symmetry check because I can not find an implementation of `all_close` for aesara sparse matrices. I only found these one https://aesara.readthedocs.io/en/latest/library/tensor/basic.html#aesara.tensor.allclose for dense.

What happens with the sampling / logp if the matrix does not respect the old requirement? Does it fail gracefully or does it return incorrect samples / logp evaluations? If it fails gracefully it should be fine to continue without the check (but then the error message should be changed since it only checks if `W` is a matrix).

aerubanov replied:
@ricardoV94 , If ```W``` not the symmetric matrix, wrong samples will be returned because it is not CAR distribution anymore. Looks not very good, so I will try to fix it somehow.

aerubanov committed:
add W symmetry check

aerubanov commented:
@ricardoV94 , I added check for ```W``` symmetry and test for it.

ricardoV94 commented on pymc3/distributions/multivariate.py:
```diff
@@ -1925,6 +1924,81 @@ def _distr_parameters_for_repr(self):
+        if not W.ndim == 2:
+            raise ValueError("W must be a matrix (ndim=2).")
+
+        sparse = isinstance(W, aesara.sparse.SparseVariable)
```
I am surprised if this works, the inputs are not evaluated inside the `make_node` are they? I think you have to add a Aesara assert instead of a Python exception

ricardoV94 replied:
There is an example here: https://github.com/pymc-devs/pymc3/blob/e8396bf8961500b63221db45b1249b92bf4afdbd/pymc3/distributions/continuous.py#L201

aerubanov replied:
@ricardoV94, I added ```Assert``` in these checks.

aerubanov committed:
add aesara assert

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -3032,41 +3032,80 @@ def test_ordered_probit_probs():
+        W = aesara.sparse.csr_from_dense(W)
+
+    car_dist = CAR.dist(mu, W, alpha, tau)
+    with pytest.raises(AssertionError):
```
Can you match the error message?

```suggestion
    with pytest.raises(AssertionError, match="W must be a symmetric adjacency matrix"):
```

And add a test for the other error (`ndim=2`)?

ricardoV94 reviewed (commented):
Looks almost there. Just suggested a tweak to the error tests

aerubanov committed:
add test for ndim check

aerubanov commented:
@ricardoV94 , I added changes that you suggested.

ricardoV94 commented:
CC @ckrapu in case he has the chance to take a look before merging

ckrapu commented:
Thanks for the ping - I am very excited to see that the fast CAR sampling method is implemented. I often find scenarios where I need to sample draws from this kind of random field even outside of the context of prior predictive sampling.

ricardoV94 reviewed (approved):
Looks good to me, trusting the `rng_fn` algorithm is correctly implemented. 

Thanks a lot @aerubanov! This was a tough one, hopefully your next PR will be easier to crack :D

twiecki commented:
Thanks @aerubanov, this was indeed a tricky one!
